lib/puppet/provider/cisco_route_map/cisco.rb

```diff
@@ -0,0 +1,588 @@
+        actions.each do |action, nu_obj|
+          rm_instances << properties_get(rmname, sequence, action, nu_obj)
+        end
+      end
+    end
+    rm_instances
+  end # self.instances
+
+  def self.prefetch(resources)
+    rm_instances = instances
+    resources.keys.each do |id|
+      provider = rm_instances.find do |rmi|
+        rmi.rmname.to_s == resources[id][:rmname].to_s &&
+        rmi.sequence.to_s == resources[id][:sequence].to_s &&
+        rmi.action.to_s == resources[id][:action].to_s
+      end
+      resources[id].provider = provider unless provider.nil?
+    end
+  end
+
+  def exists?
+    @property_hash[:ensure] == :present
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def properties_set(new_rm=false)
+    ROUTE_MAP_ALL_PROPS.each do |prop|
+      next unless @resource[prop]
+      send("#{prop}=", @resource[prop]) if new_rm
+      unless @property_flush[prop].nil?
+        @nu.send("#{prop}=", @property_flush[prop]) if
+          @nu.respond_to?("#{prop}=")
+      end
+    end
+    # custom setters which require one-shot multi-param setters
+    match_community_set
+    match_ext_community_set
+    match_ipv4_multicast_set
+    match_ipv6_multicast_set
+    match_ip_addr_access_list
+    match_ip_addr_prefix_list
+    match_route_type_set
+    set_dampening_set
+    set_distance_set
+    set_community_set
+    set_extcommunity_4bytes_set
+    set_extcommunity_rt_set
+    set_extcommunity_cost_set
+    set_ip_next_hop_set
+    set_ip_precedence
+    set_metric_set
+  end
+
+  def match_ip_addr_access_list
+    pf = @property_flush[:match_ipv4_addr_access_list]
+    v4 = pf.nil? ? @nu.match_ipv4_addr_access_list : pf
+    pf = @property_flush[:match_ipv6_addr_access_list]
+    v6 = pf.nil? ? @nu.match_ipv6_addr_access_list : pf
+    @nu.match_ip_addr_access_list(v4, v6)
+  end
+
+  def match_ip_addr_prefix_list
+    pf = @property_flush[:match_ipv4_addr_prefix_list]
+    v4 = pf.nil? ? @nu.match_ipv4_addr_prefix_list : pf
+    pf = @property_flush[:match_ipv6_addr_prefix_list]
+    v6 = pf.nil? ? @nu.match_ipv6_addr_prefix_list : pf
+    @nu.match_ip_addr_prefix_list(v4, v6)
+  end
+
+  def match_community_set
+    comm = @property_flush[:match_community] ? @property_flush[:match_community] : @nu.match_community
+    pf = @property_flush[:match_community_exact_match]
+    exact = PuppetX::Cisco::Utils.flush_boolean?(pf) ? pf : @nu.match_community_exact_match
+    @nu.match_community_set(comm, exact)
+  end
+
+  def match_ext_community_set
+    comm = @property_flush[:match_ext_community] ? @property_flush[:match_ext_community] : @nu.match_ext_community
+    pf = @property_flush[:match_ext_community_exact_match]
+    exact = PuppetX::Cisco::Utils.flush_boolean?(pf) ? pf : @nu.match_ext_community_exact_match
+    @nu.match_ext_community_set(comm, exact)
+  end
+
+  def match_ipv4_multicast_set
```

Comment:
@saichint As I was working through these examples, I setup a quick test and noticed the following issue:

I was using an Everest image so not sure if that is the problem or not.

```
Software
  BIOS: version 08.31
  NXOS: version 7.0(3)I5(2) [build 7.0(3)I5(1.224)]
  BIOS compile time:  08/26/2016
  NXOS image file is: bootflash:///CSCvc98000_nxos.7.0.3.I5.1.224.bin
  NXOS compile time:  1/31/2017 5:00:00 [02/04/2017 02:29:23]
```

1) Test manifest:
```puppet
node default {

  cisco_route_map { 'rm1 123 permit':
    match_community                          => ['public', 'private'],
    match_community_exact_match              => true,
    match_ipv4_addr_access_list              => 'access_mike',
    match_ipv4_multicast_enable              => true,
  }
}
```
2) Puppet run
```puppet
root@n9k-109#puppet agent -t 
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for n9k-109.cisco.com
Info: Applying configuration version '1486401863'
Notice: /Stage[main]/Main/Node[default]/Cisco_route_map[rm1 123 permit]/match_ipv4_multicast_enable: match_ipv4_multicast_enable changed 'false' to 'true'
Error: /Stage[main]/Main/Node[default]/Cisco_route_map[rm1 123 permit]: Could not evaluate: [#<Cisco::RouteMap:0x00000003ca6d60>] The command ' match ip multicast      ' was rejected with error:
% Incomplete command

Notice: Applied catalog in 2.01 seconds
root@n9k-109#
root@n9k-109#
```
3) With trace turned on.
```
root@n9k-109#puppet agent -t --trace
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for n9k-109.cisco.com
Info: Applying configuration version '1486401944'
Notice: /Stage[main]/Main/Node[default]/Cisco_route_map[rm1 123 permit]/match_ipv4_multicast_enable: match_ipv4_multicast_enable changed 'false' to 'true'
Error: /Stage[main]/Main/Node[default]/Cisco_route_map[rm1 123 permit]: Could not evaluate: [#<Cisco::RouteMap:0x000000044cc5a8>] The command ' match ip multicast      ' was rejected with error:
% Incomplete command

/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/cisco_node_utils-1.5.0/lib/cisco_node_utils/client/nxapi/client.rb:293:in `handle_output'
/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/cisco_node_utils-1.5.0/lib/cisco_node_utils/client/nxapi/client.rb:219:in `block in req'
/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/cisco_node_utils-1.5.0/lib/cisco_node_utils/client/nxapi/client.rb:218:in `zip'
/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/cisco_node_utils-1.5.0/lib/cisco_node_utils/client/nxapi/client.rb:218:in `req'
/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/cisco_node_utils-1.5.0/lib/cisco_node_utils/client/nxapi/client.rb:130:in `set'
/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/cisco_node_utils-1.5.0/lib/cisco_node_utils/node.rb:272:in `set'
/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/cisco_node_utils-1.5.0/lib/cisco_node_utils/node.rb:212:in `config_set'
/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/cisco_node_utils-1.5.0/lib/cisco_node_utils/node_util.rb:72:in `config_set'
/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/cisco_node_utils-1.5.0/lib/cisco_node_utils/route_map.rb:609:in `match_ipv4_multicast_set'
/opt/puppetlabs/puppet/cache/lib/puppet/provider/cisco_route_map/cisco.rb:341:in `match_ipv4_multicast_set'
/opt/puppetlabs/puppet/cache/lib/puppet/provider/cisco_route_map/cisco.rb:274:in `properties_set'
/opt/puppetlabs/puppet/cache/lib/puppet/provider/cisco_route_map/cisco.rb:585:in `flush'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type.rb:1006:in `flush'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction/resource_harness.rb:25:in `evaluate'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:230:in `apply'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:246:in `eval_resource'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:163:in `call'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:163:in `block (2 levels) in evaluate'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:386:in `block in thinmark'
/opt/puppetlabs/puppet/lib/ruby/2.1.0/benchmark.rb:294:in `realtime'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:385:in `thinmark'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:163:in `block in evaluate'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/graph/relationship_graph.rb:118:in `traverse'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:154:in `evaluate'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/catalog.rb:222:in `block in apply'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/log.rb:155:in `with_destination'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction/report.rb:142:in `as_logging_destination'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/catalog.rb:221:in `apply'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:171:in `block in apply_catalog'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:223:in `block in benchmark'
/opt/puppetlabs/puppet/lib/ruby/2.1.0/benchmark.rb:294:in `realtime'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:222:in `benchmark'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:170:in `apply_catalog'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:343:in `run_internal'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:221:in `block in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/context.rb:65:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:293:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:195:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:45:in `block (4 levels) in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent/locker.rb:21:in `lock'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:45:in `block (3 levels) in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:98:in `with_client'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:42:in `block (2 levels) in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:65:in `run_in_fork'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:41:in `block in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:179:in `call'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:179:in `controlled_run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:39:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/agent.rb:353:in `onetime'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/agent.rb:331:in `run_command'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:344:in `block in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:541:in `exit_on_fail'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:344:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:132:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:72:in `execute'
/opt/puppetlabs/puppet/bin/puppet:5:in `<main>'
Notice: Applied catalog in 1.99 seconds

```